Add support for importing using relative paths

mstemm opened the pull request:
See commits for specific changes.

I'd like feedback on the technique of modifying the ast to handle the case of potentially duplicate import statements with the same relative pathname. Not sure whether this is bad practice or not.

One open question--do we want imports of standard library modules to have a 'stdlib' prefix? Currently, the stdlib directory is in the set of search paths so one imports those files with a 'import "random" as random" rather than 'import "stdlib/random". I think I'd have to move the lib/stdlib directory somewhere as a part of this change--we don't want lib/ as a search path.

@demmer @dmajda 

This fixes #513.

dmajda commented on lib/parser/index.js:
```diff
@@ -82,7 +83,14 @@ function parse(juttle, options) {
+                let filename = module.name;
+                if (_.has(module, 'filename')) {
+                    filename = module.filename;
+                }
```
I think it would be cleaner if the protocol was that a resolver always sets `module.filename`. Then we could rely on that here.

dmajda commented on lib/parser/index.js:
```diff
@@ -97,10 +105,9 @@ function parse(juttle, options) {
     options = processOptions(options, defaultResolver);
 
-    return parseModule({ source: juttle, name: 'main' })
+    return parseModule({ source: juttle, name: 'main', filename: (options.filename ? options.filename : 'main') })
```
`options.filename` is always set in `processOptions` so it should be enough to do `filename: options.filename` here.

dmajda commented on lib/parser/index.js:
```diff
@@ -97,10 +105,9 @@ function parse(juttle, options) {
+    return parseModule({ source: juttle, name: 'main', filename: (options.filename ? options.filename : 'main') })
         .then(function() {
             asts.main.modules = buildModuleDefs(asts);
-
```
Unrelated change. (Moreover one I disagree with — in general I like to always have an empty line before a `return`.)

dmajda commented on lib/parser/index.js:
```diff
@@ -159,7 +167,15 @@ function parseSync(juttle, options) {
+        let filename = module.name;
+        if (_.has(module, 'filename')) {
+            filename = module.filename;
+        }
```
Same as above — rely on the resolver.

dmajda commented on lib/parser/index.js:
```diff
@@ -171,7 +187,7 @@ function parseSync(juttle, options) {
     options = processOptions(options, defaultResolver);
 
-    parseModule({ source: juttle, name: 'main' });
+    parseModule({ source: juttle, name: 'main', filename: options.filename ? options.filename : 'main' });
```
Same as above: `options.filename` should be enough.

dmajda commented on lib/module-resolvers/file-resolver.js:
```diff
@@ -38,36 +40,75 @@ class FileResolver {
-                    // continue looking for a module we can successfully load
+
+        if (module_path.indexOf('./') === 0 ||
+            module_path.indexOf('../') === 0) {
```
`module_path.startsWith('./') || module_path.startsWith('../')`

dmajda commented on lib/module-resolvers/file-resolver.js:
```diff
@@ -38,36 +40,75 @@ class FileResolver {
+            let dir = process.cwd();
+            if (path.isAbsolute(importer_path)) {
+                dir = path.dirname(importer_path);
+            }
```
Why is this logic needed? When, in practice, can `importer_path` be relative?

mstemm replied:
importer_path can be non-absolute when the program has a generic name like 'main'. That in turn happens for programs in the cli repl or when provided via -e.

dmajda commented on lib/module-resolvers/file-resolver.js:
```diff
@@ -38,36 +40,75 @@ class FileResolver {
+            catch (e) {
+                return false;
+            }
+        } else if (path.isAbsolute(module_path)) {
```
I think this branch could be merged with the previous one as `path.resolve` should handle an absolute path in its arguments.

dmajda commented on lib/module-resolvers/file-resolver.js:
```diff
@@ -38,36 +40,75 @@ class FileResolver {
+                    }
+                    catch (e) {
+                        // continue looking for a module we can successfully load
+                    }
```
It would be nice to split the relative module lookup and system module lookup into separate branches.

dmajda commented on docs/concepts/programming_constructs.md:
```diff
@@ -175,7 +191,7 @@ You can reference the imported symbol(s) like this:
 
 -  local_name is the local name from the import command above.
--  symbol_name is the name of the exported subgraph, constant, function, 
+-  symbol_name is the name of the exported subgraph, constant, function,
```
Put any formatting/whitespace changes into different commit(s) please.

dmajda commented:
I only reviewed the first commit as it is obvious some non-trivial changes will be needed there already.

I don’t like the `name` vs. `filename` duality and related conditions scattered around the code base at all. Basically, I think the resolver should return resolved path in the `name` property and that is what the module should be known/indexed as everywhere. Renaming this to `filename` would make it more clear, but it should be done in a separate commit as a prep change.

Overall, it would be good to split the first commit into multiple ones. At least modifying the resolver interface, adapting parser internals to handle returned module paths correctly, and teaching the file resolver to do relative imports seem like separate steps. Having them all intermingled makes the review quite hard (I went thought the changes, but I’m not sure if something wasn’t missed since the commit is not small enough to be obviously correct).

mstemm commented:
Ok I've split the work into more smaller commits and also changed the resolver interface to return the full pathname in the name: property of the returned object. The filename: property is gone. I think name: is more appropriate than filename: as filename isn't a good fit for the url resolver.

I've also addressed all the smaller changes suggested.

Please take another look?

dmajda commented on docs/concepts/programming_constructs.md:
```diff
@@ -159,10 +159,26 @@ location, then in another program you can import it as a module, using the
-relative or absolute filesystem path or a URL to a remote Juttle file containing
-exported code.
+`<module_path>` can either refer to a *User Import* or a *System
+Import*.
```
Do we really want these to be Capitalized Terms? Makes ours docs look a bit like legalese…

dmajda commented on docs/concepts/programming_constructs.md:
```diff
@@ -159,10 +159,26 @@ location, then in another program you can import it as a module, using the
+
+When resolving User Imports with relative pathnames, the pathname is
+always applied relative to the file doing the import. See the below
+examples for details.
```
“See the examples below for details.”?

dmajda commented on lib/module-resolvers/file-resolver.js:
```diff
@@ -39,22 +39,58 @@ class FileResolver {
+        if (importer_path !== undefined &&
+            path.isAbsolute(importer_path)) {
+            dir = path.dirname(importer_path);
+        }
```
Would be good to add a comment on when the importer path is non-absolute (basically your [answer here](https://github.com/juttle/juttle/pull/591#discussion_r56397960)).

dmajda commented on lib/module-resolvers/file-resolver.js:
```diff
@@ -69,7 +105,14 @@ class FileResolver {
+        var found = false;
+        if (module_path.startsWith('./') ||
+            module_path.startsWith('../') ||
+            path.isAbsolute(module_path)) {
```
Maybe put the condition into a helper method like `_is_local_import` for more clarity?

dmajda commented:
On list item is wrongly indented in the commit message of https://github.com/juttle/juttle/pull/591/commits/bd0d57fd2235261401cefe90925f63709dd77963.

dmajda commented:
For the record, I don’t like the AST modification done in https://github.com/juttle/juttle/pull/591/commits/bd0d57fd2235261401cefe90925f63709dd77963 much, but I’d say it’s good for the time being. (The whole parser-level module resolution thing needs rethinking anyway.)

dmajda commented:
+1, thanks!

> One open question--do we want imports of standard library modules to have a 'stdlib' prefix?

After thinking it through, I don’t think so. All library code will be in stdlib and the prefix would just be unnecessary bloat.

rlgomes commented:
@dmajda @mstemm @demmer  we need to pull this module resolution into `juttle-spec.js` so that the juttle spec files can import those stdlib modules without having to do "predict.juttle" as it works fine from the CLI like so: 

```rlgomes@x230> ./bin/juttle -e "import 'predict' as predict ; emit -limit 10 -from :0: | put value = count() | predict.predict"
┌────────────────────────────────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┬──────────┐
│ time                               │ value    │ E        │ P        │ S        │ T        │ Y        │ Z        │
├────────────────────────────────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┼──────────┤
│ 1970-01-01T00:00:00.000Z           │ 5        │ 0        │ 5        │ null     │ 5        │ 0        │ null     │
└────────────────────────────────────┴──────────┴──────────┴──────────┴──────────┴──────────┴──────────┴──────────┘
``` 

but if you modify one of the imports in https://github.com/juttle/juttle/tree/master/test/runtime/specs/juttle-spec/stdlib the test utilites can't resolve the standard library that way. I tried figuring out if there was a way to fix the test utilities but at the same time shouldn't this just work ? otherwise we have to also fix the module resolution within juttle-engine ?

mstemm commented:
juttle-test-utils needs to change from using its own module resolver to using the standard file resolver. Not sure why it has a custom module resolver, do you know?

rlgomes commented:
@mstemm lets wait for @dmajda he probably knows that code better than I do.

dmajda commented:
The custom resolver in `juttle-test-utils` is needed because we want the ability to specify modules inline in JuttleSpec tests (grep `### Module` to see examples) and also some JavaScript tests.

It looks like the best course of action would be to chain the inline module resolver with the file resolver and possibly the URL resolver using `resolver_utils.multiple`.
